pageserver/src/tenant/timeline.rs

```diff
@@ -3899,26 +3899,23 @@ impl Timeline {
             }
 
             // For metadata, always create delta layers.
-            let delta_layer = if !metadata_partition.parts.is_empty() {
-                assert_eq!(
-                    metadata_partition.parts.len(),
-                    1,
-                    "currently sparse keyspace should only contain a single aux file keyspace"
-                );
-                let metadata_keyspace = &metadata_partition.parts[0];
-                assert_eq!(
-                    metadata_keyspace.0.ranges.len(),
-                    1,
-                    "aux file keyspace should be a single range"
-                );
-                self.create_delta_layer(
-                    &frozen_layer,
-                    Some(metadata_keyspace.0.ranges[0].clone()),
-                    ctx,
-                )
-                .await?
+            let delta_layers = if !metadata_partition.parts.is_empty() {
+                // In the current implementation, the metadata partition will only have one part, and the part will only have
+                // one single key range. This might change in the future.
+                let mut delta_layers_created = Vec::new();
+                for ks in &metadata_partition.parts {
+                    for range in &ks.0.ranges {
+                        let layer = self
+                            .create_delta_layer(&frozen_layer, Some(range.clone()), ctx)
+                            .await?;
+                        if let Some(layer) = layer {
+                            delta_layers_created.push(layer);
+                        }
+                    }
+                }
+                delta_layers_created
```

Comment:
Does this trigger in any e2e test or just some unit test given this is under `create_image_layer` and #5863 or what appears to be recreated as #7197 is still not fixed?

Reply:
As in, if you add a panic here, does any of the regress tests fail?

Reply:
This is covered by one of the unit tests on metadata keys on its first write.

Reply:
I don't see the point of those unit tests then, if they do not show behaviour we have in prod.

Why would there be metadata keys ever in the initdb import? Are you thinking of there being metadata keys in a fullbackup import?

Reply:
I guess. Could you point me to some test which actually benefits from this patch? I don't understand how are you creating the timeline in that case.. Our usual test harness code just puts in a controlfile and that's it.

Reply:
The writer starts at initdb_lsn for https://github.com/neondatabase/neon/blob/c18b1c06460f1a55d947cdad267da4f71278655c/pageserver/src/tenant.rs#L6137-L6151 in `test_metadata_image_creation`.

Other test cases, like `test_vectored_missing_data_key_reads`, also writes metadata keys before any data keys, and I would like to refactor them using the artificial layer ingestion pull request.

Reply:
Or, to be specific, these tests do not benefit from this patch. It goes the image layer creation code path. The patch is to remove the limitations in the code path to support multiple keyspaces.

Reply:
But image layer generation should not be using this path at all, which is why I am surprised.

Aah is it because of the forced layer generation and the side-car layer building from #7899? Shouldn't be. If it was, I guess it would be strong argument against.

Reply:
Ok, re-checking the test, since we commented past each other (I didn't see your comment when I replied).

Reply:
> The writer starts at initdb_lsn for

No it starts at initdb + 0x10 (line 6138)? I don't see how even that test can trigger this. The first layer is already flushed before the image layer putting loop you linked gets to run.

> Other test cases, like test_vectored_missing_data_key_reads, also writes metadata keys before any data keys, and I would like to refactor them using the artificial layer ingestion pull request.

I don't follow. Why does it matter that metadata keys are written before other keys? To clarify: I am arguing that your code changes in this PR are no-op because in production we never take this code path, and as so, we shouldn't add test cases which pretend that it is, even though I am not seeing any test cases do that right now.

On line L3862R3862 there is:

```rust
let create_image_layer =
lsn_range.start == self.initdb_lsn && lsn_range.end == Lsn(self.initdb_lsn.0 + 1);
```

It should be the only thing which matters for this path to be taken, and it should be taken at most once, because we will have moved on with last_record_lsn and we cannot branch before the initdb_lsn. It is sort of interesting if we were able to branch at `initdb_lsn` what would/should then happen, but I think that should be unsupported.

Reply:
> Or, to be specific, these tests do not benefit from this patch. It goes the image layer creation code path. The patch is to remove the limitations in the code path to support multiple keyspaces.

Ok, I think we are now in sync.

I think we should instead add a

```
// FIXME(auxfilesv2): support multiple metadata key partitions might need initdb support as well?
```

rather than complicate the code. As you responded in some other review yesterday, there is only one partition for metadata keys **right now**. So let's not change **currently dead code** to support **currently non-existent feature**. 

I think it's not really clear if we want to even preserve the initdb image layer feature because it might only be needed in case of full backup imports. Those full backup imports could contain auxfiles, but I'd rather we'd add a regress test case for that.

If we were to fix the issue with initdb imports, we'd also need a regress test.

Reply:
Sounds good, let's drop this patch and I added a todo directly in Konstantin's pull request -> https://github.com/neondatabase/neon/pull/7099/commits/05ee51a4e07c24d967eea6ba73bdba9a68ca60e6

Reply:
> 05ee51a

Looking good!